### DISCUSSION TITLE
Feature Request: use a built-in QueryClientProvider instance when not provided

### DISCUSSION THREAD

**Discussion by KrishnaPG** [TanStack/query#4628] [Ideas]
When using frameworks such as [umi](https://umijs.org/) we do not get a chance to create the `App()` object as shown in the `QueryClientProvider` examples. (The page components are directly loaded in the layout by the framework, making it difficult to create a provider at the root level).

Besides many examples just show the creation as `const queryClient = new QueryClient()` and nothing fancy. 

It would greatly improve the Dev experience if you could make this as optional. 
  - That is, if `useQuery` does not find an active `queryClient` could you please just create one internally and use it across instead of throwing error forcing the users to create a wrapper DOM node at the root level?

Or atleast 
  - provide additional way to supply the `queryClient` instance (apart from the JSX root level wrapper `<QueryClientProvider>`). 
  
  Currently there is no way to supply a `queryClient` instance directly in the code/API apart from wrapping it through JSX wrappers, which is very inconvenient in some scenarios.

**Comment by TkDodo** [comment]
> could you please just create one internally and use it across instead of throwing error forcing the users to create a wrapper DOM node at the root level?

this is not easily possible without creating a "global variable" somewhere as a side effect

> provide additional way to supply the queryClient instance (apart from the JSX root level wrapper <QueryClientProvider>).

this will be possible with v5:
- https://github.com/TanStack/query/discussions/4252

see the point under `replace custom context with custom queryClient`

It does mean that for now, you can create a custom react context and pass that in everywhere _instead_ of wrapping in a `Provider`, but please be aware that this then changes with v5.
